src/Core/Grid/Data/Factory/AttachmentGridDataFactoryDecorator.php

```diff
@@ -0,0 +1,160 @@
+use PrestaShop\PrestaShop\Core\Grid\Record\RecordCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use PrestaShop\PrestaShop\Core\Util\File\FileSizeConverter;
+use PrestaShopBundle\Translation\TranslatorAwareTrait;
+
+/**
+ * Decorates attachment grid data factory
+ */
+final class AttachmentGridDataFactoryDecorator implements GridDataFactoryInterface
+{
+    use TranslatorAwareTrait;
+
+    /**
+     * @var GridDataFactoryInterface
+     */
+    private $attachmentDoctrineGridDataFactory;
+
+    /**
+     * @var int
+     */
+    private $employeeIdLang;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var string
+     */
+    private $dbPrefix;
+
+    /**
+     * @var FileSizeConverter
+     */
+    private $fileSizeConverter;
+
+    /**
+     * @param GridDataFactoryInterface $attachmentDoctrineGridDataFactory
+     * @param int $employeeIdLang
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param FileSizeConverter $fileSizeConverter
+     */
+    public function __construct(
+        GridDataFactoryInterface $attachmentDoctrineGridDataFactory,
+        int $employeeIdLang,
+        Connection $connection,
+        string $dbPrefix,
+        FileSizeConverter $fileSizeConverter
+    ) {
+        $this->attachmentDoctrineGridDataFactory = $attachmentDoctrineGridDataFactory;
+        $this->employeeIdLang = $employeeIdLang;
+        $this->connection = $connection;
+        $this->dbPrefix = $dbPrefix;
+        $this->fileSizeConverter = $fileSizeConverter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getData(SearchCriteriaInterface $searchCriteria)
+    {
+        $attachmentData = $this->attachmentDoctrineGridDataFactory->getData($searchCriteria);
+
+        $attachmentRecords = $this->applyModifications($attachmentData->getRecords());
+
+        return new GridData(
+            $attachmentRecords,
+            $attachmentData->getRecordsTotal(),
+            $attachmentData->getQuery()
+        );
+    }
+
+    /**
+     * @param RecordCollectionInterface $attachments
+     *
+     * @return RecordCollection
+     */
+    private function applyModifications(RecordCollectionInterface $attachments): RecordCollection
+    {
+        $modifiedAttachments = [];
+
+        foreach ($attachments as $attachment) {
+            if ((int) $attachment['products'] > 0) {
+                $productNamesArray = $this->getProductNames($attachment['id_attachment']);
+                $productNames = $productNamesArray['product_names'] ?? '';
+                $attachment['dynamic_message'] = $this->trans(
+                    'This file is associated with the following products. Are you sure you want to delete it?',
+                    [],
+                    'Admin.Catalog.Notification'
+                );
+                $attachment['dynamic_message'] .= "\n\n" . $productNames;
```

Comment:
```suggestion
                $attachment['dynamic_message'] .= PHP_EOL . PHP_EOL . $productNames;
```
Dont forget Windows users 😄

Reply:
OK now I understand why you introduce the `dynamic_message_field` property for columns ... 🤔 

It works 👍 however we'll need to document this behavior when this PR gets merged